src/dwarf_parser.h

```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+#ifdef HAVE_LIBDW
+
+#include "types.h"
+#include <elfutils/libdwfl.h>
+#include <map>
+#include <set>
+#include <string>
+
+namespace bpftrace {
+
+class Dwarf
+{
+public:
+  Dwarf(const std::string &file_path);
+  virtual ~Dwarf();
+
+  std::vector<std::string> get_function_params(
+      const std::string &function) const;
+
+  bool has_data;
+
+private:
+  std::map<std::string, Dwarf_Die> function_param_dies(
+      const std::string &function) const;
+
+  std::optional<Dwarf_Die> get_func_die(const std::string &function) const;
+
+  std::string get_type_name(Dwarf_Die &type_die) const;
+
+  Dwfl *dwfl = nullptr;
+  std::string file_path;
+};
+
+} // namespace bpftrace
+
+#else
+
+#include "log.h"
+
+namespace bpftrace {
+
+class Dwarf
+{
+public:
+  Dwarf(const std::string &file_path __attribute__((unused)))
+  {
+    if (!warned)
+      LOG(WARNING) << "Cannot parse DWARF: libdw not available";
+    warned = true;
+  }
+
+  std::vector<std::string> get_function_params(const std::string &function
+                                               __attribute__((unused))) const
+  {
+    return {};
+  }
+
+  std::map<std::string, SizedType> resolve_function_params(
+      const std::string &function __attribute__((unused))) const
+  {
+    return {};
+  }
+
+  bool has_data = false;
+
+private:
+  static bool warned;
+};
+
+bool Dwarf::warned = false;
```

Comment:
Not sure you can do this in a header. Might be an ODR violation. Move to .cpp instead?

See https://stackoverflow.com/questions/185844/how-to-initialize-private-static-members-in-c